# Create baseline for DKT

## Clone knowledge-tracing repo

In [ ]:
!git clone https://github.com/aarushban-12/knowledge-tracing-collection-pytorch.git
%cd knowledge-tracing-collection-pytorch

!pip install -r requirements.txt

Cloning into 'knowledge-tracing-collection-pytorch'...
remote: Enumerating objects: 488, done.
remote: Counting objects: 100% (47/47), done.
remote: Compressing objects: 100% (45/45), done.
remote: Total 488 (delta 22), reused 4 (delta 0), pack-reused 441 (from 1)
Receiving objects: 100% (488/488), 1.11 MiB | 5.79 MiB/s, done.
Resolving deltas: 100% (252/252), done.
/content/knowledge-tracing-collection-pytorch
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.2/17.2 MB 44.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 915.3/915.3 kB 39.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.5/77.5 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.8/59.8 kB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 51.7 MB/s eta 0:00:00


## Import dataset and fix encoding issue

In [ ]:
!mkdir -p datasets/ASSIST2009
!find datasets -maxdepth 2 -type d

from google.colab import drive
drive.mount('/content/drive')
!cp "/content/drive/MyDrive/education-ml-research/ASSISTments2009/skill_builder_data.csv" \
    "datasets/ASSIST2009/skill_builder_data.csv"

import pandas as pd

path = "datasets/ASSIST2009/skill_builder_data.csv"

## Dataset only works in latin1, so first read in latin1 then convert to utf-8 to be used by SAKT
df = pd.read_csv(path, encoding="latin-1")

print(df.shape)
print(df.columns.tolist())

df.to_csv(
    "datasets/ASSIST2009/skill_builder_data_utf8.csv",
    index=False,
    encoding="utf-8"
)

!mv datasets/ASSIST2009/skill_builder_data.csv datasets/ASSIST2009/skill_builder_data_original.csv
!mv datasets/ASSIST2009/skill_builder_data_utf8.csv datasets/ASSIST2009/skill_builder_data.csv

pd.read_csv(
    "datasets/ASSIST2009/skill_builder_data.csv"
).head()

datasets
datasets/ASSIST2009
Mounted at /content/drive


/tmp/ipykernel_843/3009879563.py:13: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(path, encoding="latin-1")


(525534, 30)
['order_id', 'assignment_id', 'user_id', 'assistment_id', 'problem_id', 'original', 'correct', 'attempt_count', 'ms_first_response', 'tutor_mode', 'answer_type', 'sequence_id', 'student_class_id', 'position', 'type', 'base_sequence_id', 'skill_id', 'skill_name', 'teacher_id', 'school_id', 'hint_count', 'hint_total', 'overlap_time', 'template_id', 'answer_id', 'answer_text', 'first_action', 'bottom_hint', 'opportunity', 'opportunity_original']


/tmp/ipykernel_843/3009879563.py:27: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv(


,order_id,assignment_id,user_id,assistment_id,problem_id,original,correct,attempt_count,ms_first_response,tutor_mode,...,hint_count,hint_total,overlap_time,template_id,answer_id,answer_text,first_action,bottom_hint,opportunity,opportunity_original
0,33022537,277618,64525,33139,51424,1,1,1,32454,tutor,...,0,3,32454,30799,NaN,26,0,NaN,1,1.0
1,33022709,277618,64525,33150,51435,1,1,1,4922,tutor,...,0,3,4922,30799,NaN,55,0,NaN,2,2.0
2,35450204,220674,70363,33159,51444,1,0,2,25390,tutor,...,0,3,42000,30799,NaN,88,0,NaN,1,1.0
3,35450295,220674,70363,33110,51395,1,1,1,4859,tutor,...,0,3,4859,30059,NaN,41,0,NaN,2,2.0
4,35450311,220674,70363,33196,51481,1,0,14,19813,tutor,...,3,4,124564,30060,NaN,65,0,0.0,3,3.0


## Train the model

In [ ]:
!python train.py --model_name=dkt --dataset_name=ASSIST2009

/content/knowledge-tracing-collection-pytorch/data_loaders/assist2009.py:58: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(self.dataset_path).dropna(subset=["skill_name"])\
/content/knowledge-tracing-collection-pytorch/models/utils.py:91: UserWarning: The torch.cuda.*DtypeTensor constructors are no longer recommended. It's best to use methods such as torch.tensor(data, dtype=*, device='cuda') to create tensors. (Triggered internally at /pytorch/torch/csrc/tensor/python_tensor.cpp:78.)
  q_seqs.append(FloatTensor(q_seq[:-1]))
Epoch: 1,   AUC: 0.7631532172447847,   Loss Mean: 0.6430513262748718
Epoch: 2,   AUC: 0.7918418235609437,   Loss Mean: 0.6089233756065369
Epoch: 3,   AUC: 0.8020579739220327,   Loss Mean: 0.5986979603767395
Epoch: 4,   AUC: 0.8061784809609771,   Loss Mean: 0.5934377312660217
Epoch: 5,   AUC: 0.8099994163916813,   Loss Mean: 0.5895358920097351
Epoch: 6,   AUC: 0.8122829661867657,   Loss Mean: 

## AUC data

In [ ]:
!find . -name "aucs.pkl"

import pickle
import numpy as np

auc_path = "./ckpts/dkt/ASSIST2009/aucs.pkl"

with open(auc_path, "rb") as f:
    aucs = pickle.load(f)

print("Number of epochs:", len(aucs))
print("Best AUC:", max(aucs))
print("Best epoch:", np.argmax(aucs) + 1)
print("Final AUC:", aucs[-1])

./ckpts/dkt/ASSIST2009/aucs.pkl
Number of epochs: 100
Best AUC: 0.8201386933473436
Best epoch: 19
Final AUC: 0.7728097306446675


## Model Comparison

| Model | Dataset | Key Settings | Test AUC |
|:---|:---|:---|---:|
| **SAKT** | ASSISTments 2009 | Train ratio = 0.8, batch size = 128 | **0.8098** |
| **BKT** | ASSISTments 2009 | 80/20 random split, 123 skills, C++ backend | **0.7945** |
| **DKT** | ASSISTments 2009 | Train ratio = 0.8, batch size = 128, 100 epochs | **0.8201** |
| **Logistic Regression** | ASSISTments 2009 | Student accuracy, skill accuracy, previous skill attempts, time since last attempt | **0.7212** |

## Comparison with Published Results

My DKT implementation achieved a best test AUC of **0.8201** on
ASSISTments 2009.

This result is very close to the **0.8215 ± 0.0005** AUC reported by
HCNOH for DKT on ASSISTments 2009. However, the results are not
directly identical experiments because the HCNOH repository uses
different training configurations, including a 90/10 train-test
ratio, whereas this experiment uses an 80/20 student-level split.

The original DKT paper by Piech et al. reported an AUC of **0.86**
on ASSISTments. The difference between this result and our result
likely reflects differences in data preprocessing, train-test
splitting, model configuration, and training procedure. Therefore,
the published AUC is used as a reference point rather than as a
directly controlled benchmark.

Overall, the close agreement between our result and the HCNOH
implementation provides evidence that our DKT baseline is
implemented and trained correctly.